lib/ngtsc/symbol.ts

```diff
@@ -0,0 +1,87 @@
+import { WorkspaceSymbols } from './workspace.symbols';
+import { ClassDeclaration, Decorator } from '@angular/compiler-cli/src/ngtsc/reflection';
+import { Declaration } from 'typescript';
+import { TraitState, Trait } from '@angular/compiler-cli/src/ngtsc/transform';
+import { R3DependencyMetadata } from '@angular/compiler';
+import type { AnnotationNames } from './utils';
+import type { FactoryOutput } from './find-symbol';
+
+const handlerName = {
+  'NgModule': 'NgModuleDecoratorHandler',
+  'Pipe': 'PipeDecoratorHandler',
+  'Injectable': 'InjectableDecoratorHandler',
+  'Directive': 'DirectiveDecoratorHandler',
+  'Component': 'ComponentDecoratorHandler'
+} as const;
+
+const filterByHandler = (annotation: AnnotationNames) => (trait: Trait<Decorator, any, unknown>) => {
+  return trait.handler.name === handlerName[annotation];
+};
+
+
+export abstract class Symbol<AnalysisData> {
+  protected readonly abstract annotation: AnnotationNames;
+  protected readonly abstract deps: R3DependencyMetadata[] | 'invalid' | null;
+  private _trait?: Trait<Decorator, AnalysisData, unknown>;
```

Comment:
We always have this defined thanks to the getter, right? We can make it non-optional.

Reply:
The `record` might be undefined of the node is not analyzed. I'll change the type to `Trait<...> | undefined` to make it more obvious

Reply:
Since in the [getter we check if `_trait`](https://github.com/ng-ast/ngast/pull/23/files/a0359e98cc0da7bca34f76beead42ce0e1575e40#diff-ef10726d20d9e9e8a12afa43ac45c779R64) is initialized and if it isn't we initialize it, what's the case when `_trait` could be undefined?

Reply:
If there is no record in the TraitCompiler for this ClassDeclaration the `record` will be undefined, so there is no `_trait` attached to it. This can happen if : 
- The class is outside of the scope of the workspace (since one can instanciate this class on **any** ClassDeclaration)
- The analysis has not been done yet.